File: Projects/SwiftPerfLab/Models/GitHubEvent.swift
import Foundation

/// The root Data Transfer Object (DTO) for GitHub telemetry parsing.
///
/// Conforms strictly to `Decodable` instead of `Codable` to prevent the compiler from 
/// generating unused serialization metadata, reducing binary size and compilation overhead. 
/// It is marked as `Sendable` to guarantee thread-safe transitions from background parsing tasks to the UI.
struct GitHubEvent: Decodable, Identifiable, Sendable {
    
    /// A unique identifier sourced directly from the raw JSON payload.
    /// Bypasses local `UUID()` generation to eliminate CPU overhead during benchmark loops.
    let id: String
    let type: String
    let actor: Actor
    let repo: Repo
    let payload: Payload
    let publicEvent: Bool
    
    /// Stored as a raw String rather than a Date object.
    /// ISO8601 date parsing introduces computational overhead. Using a String ensures the benchmark 
    /// measures pure JSON structural mapping throughput without date-formatting delays.
    let createdAt: String

    /// Defines the exact mapping between snake_case JSON keys and camelCase properties.
    /// Explicitly defining CodingKeys provides O(1) compile-time mapping, which is significantly 
    /// faster than using `.convertFromSnakeCase` at runtime.
    enum CodingKeys: String, CodingKey {
        case id, type, actor, repo, payload
        case publicEvent = "public"
        case createdAt = "created_at"
    }
}

/// Represents the entity (user or bot) that triggered the GitHub event.
struct Actor: Decodable, Sendable {
    let id: Int
    let login: String
    let url: String
    let avatarUrl: String

    enum CodingKeys: String, CodingKey {
        case id, login, url
        case avatarUrl = "avatar_url"
    }
}

/// Represents the target repository associated with the event.
struct Repo: Decodable, Sendable {
    let id: Int
    let name: String
    let url: String
}

/// Encapsulates event-specific metadata payloads.
/// Properties are optional to safely handle incomplete payload structures without throwing decoder errors.
struct Payload: Decodable, Sendable {
    let ref: String?
    let refType: String?
    let masterBranch: String?
    let description: String?

    enum CodingKeys: String, CodingKey {
        case ref
        case refType = "ref_type"
        case masterBranch = "master_branch"
        case description
    }
}

File: Projects/KotlinMultiPlatformPerfLab/composeApp/src/commonMain/kotlin/com/tez/perflab/models/GitHubEvent.kt
package com.tez.perflab.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * A cross-platform Data Transfer Object (DTO) modeling the GitHub Events API.
 *
 * Utilizing the `@Serializable` annotation triggers compile-time code generation.
 * This avoids the CPU overhead introduced by JVM reflection, improving parsing throughput.
 */
@Serializable
data class GitHubEvent(
    val id: String,
    val type: String,
    val actor: Actor,
    val repo: Repo,
    val payload: Payload,

    // Maps the JSON key "public" to a Kotlin-safe property name to avoid keyword collisions.
    @SerialName("public")
    val publicEvent: Boolean,

    @SerialName("created_at")
    val createdAt: String
)

@Serializable
data class Actor(
    val id: Int,
    val login: String,
    val url: String,

    @SerialName("avatar_url")
    val avatarUrl: String
)

@Serializable
data class Repo(
    val id: Int,
    val name: String,
    val url: String
)

/**
 * Models the polymorphic 'payload' block of a GitHub event structure.
 *
 * Optional parameters are used to prevent mapping crashes. Since payload schemas
 * vary based on the event type, this flat optional structure avoids the CPU penalties
 * of custom polymorphic deserializers during high-frequency benchmarking.
 */
@Serializable
data class Payload(
    val ref: String? = null,

    @SerialName("ref_type")
    val refType: String? = null,

    @SerialName("master_branch")
    val masterBranch: String? = null,

    val description: String? = null
)